### DISCUSSION TITLE
Oqtane 4.0.0 Enhancements

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2848] [General]
The #2826 post from last week provided a lot of context in regards to the challenges of managing page resources such as JavaScript, CSS, and Meta tags in a Blazor application. I am happy to report that all of the previously described scenarios are now fully supported.

In regards to Title, CSS and Meta tags, the new Head component capability works well in both pre-rendering and interactive scenarios... eliminating the need for complicated and costly JS Interop (which was used in prior versions of Oqtane to maintain these page elements). Administrators now have the ability to specify the Title, CSS and Meta tags at both the Head and Body level as well as at the Site and Page level, and developers have the ability to utilize new ModuleBase methods for SetPageTitle() and AddHeadContent() methods to programmatically manage these elements. 

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/8047ab33-583d-4d9b-a1d5-96dc50bf5335)

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/5d7b9cc7-94b9-4c13-806e-42bada45fad8)

JavaScript is an entirely different challenge in Blazor and can not be supported using a standard Razor component approach (in fact Microsoft provides explicit guidance "Don't place a script tag in a Razor component file because the script tag can't be updated dynamically by Blazor"). Therefore, Oqtane has to rely on a custom approach. 

First, it is important to recognize that there are a variety of different JavaScript registration requirements in Oqtane - sometimes they need to be injected at the Site level, sometimes at the Page/Theme level, and sometimes at the Module level. It is also important to note that there are only 2 specific locations in a Blazor application where JavaScript can be safely injected - during server rendering (in _host.cshtml) and on the client in OnAfterRenderAsync() (using JS Interop). There are also the differences in JavaScript formats to take into consideration ie. inline vs. external. And differences in JavaScript page locations ie. head vs. body. So it was important to map out all of the different use cases so that each could be handled appropriately. 

The end result is that Site level JavaScript is always injected during rendering on the server (in _host.cshtml). Page/Theme level and Module level are injected on the client in OnAfterRenderAsync() (using JS Interop). The JS Interop approach continues to leverage a tiny dependency management library named LoadJS to manage the loading of external JavaScript files. And for inline scripts the native Interop method has been enhanced to remove the script from the DOM if it already exists (based on Id) so that it can be re-executed. 

In addition to the changes outlined above, it is now possible to specify your CSS and JavaScript resources in the IModule and ITheme interfaces - eliminating the need to duplicate these references in each module component. When using this approach it is now also possible to use the Level property to specify that a JavaScript resource is ResourceLevel.Site which indicates that it should be injected during server rendering. This improvement effectively eliminates the need for the legacy IHostResources, so the interface will be deprecated and no longer be supported.

```
namespace Oqtane.Modules.HtmlText
{
    public class ModuleInfo : IModule
    {
        public ModuleDefinition ModuleDefinition => new ModuleDefinition
        {
            Name = "HtmlText",
            Description = "Renders HTML or Text Content",
            Version = "1.0.1",
            ServerManagerType = "Oqtane.Modules.HtmlText.Manager.HtmlTextManager, Oqtane.Server",
            ReleaseVersions = "1.0.0,1.0.1",
            SettingsType = "Oqtane.Modules.HtmlText.Settings, Oqtane.Client",
            Resources = new List<Resource>()
            {
                new Resource { ResourceType = ResourceType.Script, Level = ResourceLevel.Site, Url = "~/Module.js" }
            }
        };
    }
}
```

You will notice the addition of the ``~`` syntax as a static identifier which will be replaced dynamically at runtime with the appropriate local path. This was necessary because helper methods such as ModulePath() and ThemePath() are not available in this scenario - however it is important to note that the ``~`` syntax can now also be used in component resource declarations as well.

In order to support Site level resources, it was necessary to be able to identify the modules and themes which are enabled for a specific Site in an installation. So an enhancement was added which allows a Host user to indicate if a Module or Theme is enabled in a specific site. The default behavior is that modules and themes are always enabled by default, however it can also be controlled using a new property of IsAutoEnabled which can be set in the IModule or ITheme interface. 

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4cb5c4b7-a08f-4739-8254-811681a860ef)

For Themes, the ability to be able to specify if they are enabled or not meant that they needed to be managed in the database like other entities (previously they were only managed in memory at run-time). Treating themes as a database entity meant that the standard audit information is now retained, and it is also possible to allow a Host user to modify the theme name. And from a development perspective it means that themes can also have settings just like any other Oqtane entity.

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/afde4146-362e-4d7d-b20d-5ec7c510c0f1)

Now that modules and themes can be enabled on a per site basis, it is also possible to optimize the client assembly loading behavior for WebAssembly and Hybrid hosting models. In prior versions, the assembly loading relied on reflection/interfaces and downloaded all Oqtane client assemblies which existed in the entire installation. The new method eliminates the duplicate reflection logic and downloads only those assemblies which are enabled for the current site. 

In addition to all of the page resource enhancements, there were a number of other improvements added this past week...

Based on a feature request from @thabaum, a validation capability was added to user profiles. So it is now possible to specify a regular expression (RegExp) as a validation requirement for any profile field.

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/a926b090-3d02-4705-9ceb-9933eca6eb87)

And based on a feature request from @leighpointer, the ability to disable SMTP for a specific site was added. This is helpful in scenarios where you want to retain the SMTP configuration you have entered previously in the system but want to disable the service from delivering notifications.

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/5fa4b896-5cec-49f4-b9e3-d6e5bd8b26b9)

**Comment by sbwalker** [comment]
The next feature I am working on is primarily for developers and is intended to automate the configuration of a module in a site. ..

For those of you familiar with Blazor it has the concept of "routable components" and "standard components". Routable components are identical to standard components except that they use an ``@page`` attribute to indicate a specific route that can be used to render the component in the UI. The ``@page`` attribute is used by the default Blazor router to create a route table  to enable this functionality. However based on the fact that Oqtane uses its own custom router and utilizes a purely dynamic rendering model, "routable components" are currently not supported in Oqtane.

One of the challenges which many developers face in Oqtane is how to automate the configuration of their site with their custom modules. Obviously there is an administrative UI which can be used to perform this task manually, however many developers prefer the consistency and predictability of static configuration. To address this need Oqtane supports an ISiteTemplate capability which allows developers to create custom site templates which can be used when creating a new site to provision pages, modules, content, etc... It also has an ISiteMigration capability which allows you to programmatically specify version-based migration logic to manipulate the configuration of your site as it evolves over time. The combination of these 2 capabilities provides some excellent opportunities for automation so that developers can avoid manual configuration. However, based on nuances of the application life cycle and where these methods need to be executed, they can be rather complex for some developers to grasp. So it seems that a simpler method could also provide some benefits.

The idea is to utilize the concept of "routable components" within the Oqtane environment. I like to refer to this as "routable modules" and the goal would be to provide a declarative method where a module developer can specify the configuration details of how their module should be integrated into a site. Basically I am envisioning another property on the IModule interface where a developer could define the site, page name, appearance characteristics, permissions, module title, etc... This static definition would be utilized at startup to automatically provision the database with all of the specified configuration details. And the result would be that when the site loads, the UI would be fully configured with pages and modules. In addition, if the configuration details changed in the future, these changes could also be reconciled at startup... providing a perpetual mechanism to automate site configuration. Obviously, this feature would be limited to page and module properties - so if there was additional configurations required they could still be handled through ISiteMigration.

**Reply by vnetonline** [reply]
@sbwalker have the theme and module templates been updated to use the new way of attaching resources for head and body ?

**Reply by sbwalker** [reply]
No, the default templates have not been updated yet... please go ahead and update them if you have time. Note that the Resources should not be specified as Site level in the templates.

**Comment by sbwalker** [comment]
So... when running in Blazor Hybrid/MAUI there is no server-side *.cshtml page and no pre-rendering support... which means that the only way to include JavaScript in the index.html file is to inject it dynamically. So basically this means that the framework will need to detect if the client is Blazor Hybrid/MAUI and utilize JavaScript injection in those scenarios (ie. there will be some differences in run-time behavior based on the environment).

**Comment by iJungleboy** [comment]
May I say that Christmas is certainly early in 2023!!! 🙏🏽 💪🏽

**Comment by nicpitsch** [comment]
Strange, the **description** meta tag in *Head Content* - like `<meta name="description" content="Mozilla is the not-for-profit behind the lightning fast Firefox browser. We put people over profit to give everyone more power online.">` - won't be set. That is I can't see it in page source.
Other meta tag's like *robots*, *google-site-verification*, *viewport*, *keywords* (deprecated by google) however appear in page source.

**Comment by sbwalker** [comment]
@nicpitsch see #3102 (thank you for reporting)
